contracts/JoeBarV2.sol

```diff
@@ -0,0 +1,135 @@
+// SPDX-License-Identifier: MIT
+
+/// SPDX-License-Identifier: MIT
+
+pragma solidity 0.6.12;
+
+import "@openzeppelin/contracts-upgradeable/token/ERC20/IERC20Upgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/math/SafeMathUpgradeable.sol";
+
+import "@openzeppelin/contracts-upgradeable/proxy/Initializable.sol";
+import "@openzeppelin/contracts-upgradeable/token/ERC20/ERC20Upgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+
+interface IBarRewarder {
+    function initialize(address _bar) external;
+
+    function claimReward() external;
+}
+
+/**
+ * @title Joe Bar V2
+ * @author traderjoexyz
+ * @notice JoeBar is the coolest bar in town. You come in with some JOE, and leave with more!
+ * The longer you stay, the more JOE you get. This contract handles swapping to and from moJOE,
+ * One of TraderJoe's staking token.
+ * @dev All public/external functions are tested in `./test/JoeBarV2.test.ts`.
+ */
+contract JoeBarV2 is Initializable, ERC20Upgradeable, OwnableUpgradeable {
+    using SafeMathUpgradeable for uint256;
+
+    /// @notice JOE Token address
+    IERC20Upgradeable public joe;
+
+    /*
+     * @notice Rewarder address
+     * @dev Needs to be set up after deploying barV2
+     */
+    IBarRewarder public rewarder;
+
+    /// @notice entry fee, in basis points aka parts per 10,000 so 5000 is 50%, entry fee of 50%.
+    uint256 public entryFee;
+
+    /// @notice Emitted when owner set the entry fee.
+    event SetEntryFee(uint256 _entryFee);
+
+    /// @notice Emitted when owner set the rewarder.
+    event SetRewarder(address _rewarder);
+
+    /// @notice Defines the moJOE token contract
+    function initialize(
+        address _joe,
+        address _rewarder,
+        uint256 _entryFee
+    ) public initializer {
```

Comment:
nit: make `external`

Reply:
Constructor must be public or internal. It can't be external